contracts/TokenRules.sol

```diff
@@ -64,17 +64,29 @@ contract TokenRules {
         string ruleAbi;
     }
 
+    /**
+     * RuleIndex struct is going to be used in 'rulesByAddress' and
+     * 'rulesByNameHash' mappings for pointing to the index with 'rules' array.
+     * Simple usage of uint256 in those mappings does not work, because
+     * for non existing rule name and address it defaults to 0 index,
+     * which is obviously wrong. Before accessing 'rules' array by index
+     * one should check 'exists' field of the struct.
+     */
+    struct RuleIndex {
+        uint256 index;
+        bool exists;
+    }
```

Comment:
Below approach can avoid extra storage:

We can iterate on rules array and check if input ruleNameHash or ruleAddress exist.
The method name can be findRule similar to findGlobalConstraintIndex.
registerRule is not a frequent operation by organization. So the extra iteration can work.

Let me know your opinion on this.

Reply:
Traversing the array is not a good idea and should be avoided if possible. It costs gas and in case of many rules could fail because of gas limit.

Reply:
ok. I also think we should also follow the same bool exists approach for searching globalConstraints as same logic applies there also. Until there is a strong reason we should do it. Not a work for this PR but something we can take up as a separate issue. Sorry for not commenting that in previous PR reviews.